hack/generate_tests.py

```diff
@@ -0,0 +1,157 @@
+#!/usr/bin/python
+"""Regenerate tests for only the files that have changed."""
+
+import argparse
+import logging
+import os
+import subprocess
+
+TOP_LEVEL_EXCLUDES = ["docs"]
+
+def generate_test_name(repo_root, package_dir):
+  """Generate the name of the go file to write the test to.
+
+  Args:
+    repo_root: Root of the repository
+    package_dir: Full path to the kustomize directory.
+  """
+
+  rpath = os.path.relpath(package_dir, repo_root)
+
+  pieces = rpath.split(os.path.sep)
+  name = "-".join(pieces) + "_test.go"
+
+  return name
+
+def get_changed_dirs():
+  """Return a list of directories of changed kustomization packages."""
+  # Generate a list of the files which have changed with respect to the upstream
+  # branch
+  modified_files = subprocess.check_output(
+    ["git", "diff", "--name-only", "@{upstream}"])
+
+  repo_root = subprocess.check_output(["git", "rev-parse", "--show-toplevel"])
+  repo_root = repo_root.decode()
+  repo_root = repo_root.strip()
+
+  modified_files = modified_files.decode().split()
+  changed_dirs = set()
+  for f in modified_files:
+    changed_dir = os.path.dirname(os.path.join(repo_root, f))
+    # Check if its a kustomization directory by looking for a kustomization
+    # file
+    if not os.path.exists(os.path.join(repo_root, changed_dir,
+                                       "kustomization.yaml")):
+      continue
+
+    changed_dirs.add(changed_dir)
+
+    # If the base package changes then we need to regenerate all the overlays
+    # We assume that the base package is named "base".
+
+    if os.path.basename(changed_dir) == "base":
+      parent_dir = os.path.dirname(changed_dir)
+      for root, _, files in os.walk(parent_dir):
+        for f in files:
+          if f == "kustomization.yaml":
+            changed_dirs.add(root)
+
+  return changed_dirs
+
+def find_kustomize_dirs(root):
+  """Find all kustomization directories in root"""
+
+  changed_dirs = set()
+  for top in os.listdir(root):
+    if top.startswith("docs"):
+      print("donotsubmit")
```

Comment:
Could we log instead?

Reply:
Thanks for catching this. This is debug code I'll remove in my next commit.

side bar: I'm really hoping there is a github action that we can use to automatically catch donotsubmit in code.

Reply:
This would be an interesting integration for GitHub.